src/dsql/StmtNodes.cpp

```diff
@@ -5632,6 +5689,16 @@ StmtNode* ForNode::pass2(thread_db* tdbb, CompilerScratch* csb)
 	if (rse->hasWriteLock())
 		withLock = true;
 
+	if (rse->hasSkipLocked())
+	{
+		SortedStreamList streamList;
+		rse->collectStreams(streamList);
+		for (auto stream : streamList)
+		{
+			csb->csb_rpt[stream].csb_flags |= csb_skip_locked;
+		}
+	}
+
```

Comment:
IIRC, `rse->collectStreams()` collects all the streams found inside the RSE, including those buried inside subqueries inside different expressions in FIRST/SKIP/WHERE/etc.

I'd move this code to `Optimizer::compile()` instead, it already has `compileStreams` array which contains only streams inside the `FROM` clause. Maybe even share the existing loop between `WITH LOCK` and `SKIP LOCKED` features.